Fast browse mode - "browse mode moves system focus off" mode

mltony opened the pull request:
### Link to issue number:
https://github.com/nvaccess/nvda/issues/2039

### Summary of the issue:
NVDA has systematic problems with browse mode in all browsers, such as:
* Rubber band effect: NVDA sporadically undoes the last browse mode keystroke by jumping to the previous location
* Edit boxes steal system focus when arrowing down through them on some websites
* Browse mode keystrokes are slow to respond. 
Examples of websites where users have reported significant problems are Facebook and GMail among others.
All these problems are especially bothersome on large webpages with a lot of JavaScript code and form elements. Jaws doesn't seem to have these problems in browse mode.
### Description of how this pull request fixes the issue:
This PR introduces "Browse mode moves system focus off" mode that can be toggled on and off by NVDA+8 keystroke. This new mode is disabled by default, so it is unlikely it will cause any user complaints when this mode is turned off. When system focus doesn't follow browse mode focus, NVDA doesn't send .setFocus() and .scrollIntoView() requests to underlying browsers on every keystroke. It will however send setFocus() event when user interacts with a page element. This appears to have solved all three symptoms described above.
This PR also intercepts new keystrokes, such as control+Enter and applications key, as it needs to realign  system focus with browse mode focus in these cases. After setting system focus it still passes the gesture to the operating system, so there is no change in user experience.
### Testing performed:
I have tested this PR in all 4 browsers and didn't notice any side effects. This PR seems to have solved all three symptoms I described above. Browsing experience remains unchanged - no new keystrokes to learn, except for improvement in the symptoms described above.
Only tested on Windows 10.
### Known issues with pull request:
No known issues known at this time.
I realize that this mode might affect the way NVDA interacts with many applications, that's why I made it as a mode that can be turned on and off. To claim this new mode is stable more extensive testing by the users will be needed.
### Change log entry:
"Browse mode moves system focus off" mode added to improve certain problems with browse mode.
Section: New features

mltony committed:
Fast browse mode - system focus doesn't follow browse mode focus

josephsl commented:
CC @LeonardDer regarding possible conflicts with vision framework, @DerekRiemer and @JCSTeh for overall analysis on browse mode facilities please.

 

From: mltony <notifications@github.com> 
Sent: Monday, December 31, 2018 4:03 PM
To: nvaccess/nvda <nvda@noreply.github.com>
Cc: Subscribed <subscribed@noreply.github.com>
Subject: [nvaccess/nvda] Fast browse mode - "browse mode moves system focus off" mode (#9114)

 


Link to issue number:


#2039 <https://github.com/nvaccess/nvda/issues/2039> 


Summary of the issue:


NVDA has systematic problems with browse mode in all browsers, such as:

*	Rubber band effect: NVDA sporadically undoes the last browse mode keystroke by jumping to the previous location
*	Edit boxes steal system focus when arrowing down through them on some websites
*	Browse mode keystrokes are slow to respond.
Examples of websites where users have reported significant problems are Facebook and GMail among others.
All these problems are especially bothersome on large webpages with a lot of JavaScript code and form elements. Jaws doesn't seem to have these problems in browse mode.


Description of how this pull request fixes the issue:


This PR introduces "Browse mode moves system focus off" mode that can be toggled on and off by NVDA+8 keystroke. This new mode is disabled by default, so it is unlikely it will cause any user complaints when this mode is turned off. When system focus doesn't follow browse mode focus, NVDA doesn't send .setFocus() and .scrollIntoView() requests to underlying browsers on every keystroke. It will however send setFocus() event when user interacts with a page element. This appears to have solved all three symptoms described above.
This PR also intercepts new keystrokes, such as control+Enter and applications key, as it needs to realign system focus with browse mode focus in these cases. After setting system focus it still passes the gesture to the operating system, so there is no change in user experience.


Testing performed:


I have tested this PR in all 4 browsers and didn't notice any side effects. This PR seems to have solved all three symptoms I described above. Browsing experience remains unchanged - no new keystrokes to learn, except for improvement in the symptoms described above.
Only tested on Windows 10.


Known issues with pull request:


No known issues known at this time.
I realize that this mode might affect the way NVDA interacts with many applications, that's why I made it as a mode that can be turned on and off. To claim this new mode is stable more extensive testing by the users will be needed.


Change log entry:


"Browse mode moves system focus off" mode added to improve certain problems with browse mode.
Section: New features

  _____  


You can view, comment on, or merge this pull request online at:


  https://github.com/nvaccess/nvda/pull/9114


Commit Summary


*	Fast browse mode - system focus doesn't follow browse mode focus


File Changes


*	M source/browseMode.py <https://github.com/nvaccess/nvda/pull/9114/files#diff-0>  (45) 
*	M source/config/configSpec.py <https://github.com/nvaccess/nvda/pull/9114/files#diff-1>  (1) 
*	M source/globalCommands.py <https://github.com/nvaccess/nvda/pull/9114/files#diff-2>  (15) 
*	M source/gui/settingsDialogs.py <https://github.com/nvaccess/nvda/pull/9114/files#diff-3>  (6) 


Patch Links:


*	https://github.com/nvaccess/nvda/pull/9114.patch
*	https://github.com/nvaccess/nvda/pull/9114.diff

—
You are receiving this because you are subscribed to this thread.
Reply to this email directly, view it on GitHub <https://github.com/nvaccess/nvda/pull/9114> , or mute the thread <https://github.com/notifications/unsubscribe-auth/AHgLkNF4103CyWjWpsDxs5MpW2QeXXpyks5u-qWYgaJpZM4Zltdg> .

josephsl reviewed (changes requested):
A few general comments:

Consider the following scenario: a user opens something other than web browsers (say, an audio editing program), and the user presses NVDA+number row 8.

* What will the user hear?
* What should the user hear?
* What will you hear?
* What should you hear?
* If what users should hear versus what users actually hear are different, what do you think might be happening, and are there possible solutions?

I think it would be helpful to detect non-browse mode scenarios and either:

* Present an error message.
* Do not allow this to be configurable via a gesture.

I'd like to recommend the second solution in line with screen layout command (NVDA+V).

Next, regarding the key string in use, the way other NVDA settings do it is:

settingName

So for this one, the camel-case name would be "focusFollowsBrowse" (without quotes).

Lastly, instead of providing gestures and input help messages for commonly encountered commands where you may need to sync system focus (or rather, individual scripts for each command), I recommend following the script binding trick used in places such as Outlook app module where you have one single script (no input help message) with some commonly encountered commands assigned to it via bindGesture function call. This whole thing may change if you find a more efficient solution (Jamie is our go-to person on browse mode).

Thanks.

mltony commented:
@josephsl ,
1. Regarding what should happen if the keystroke is pressed not in browser. I don't htink a strong opinion either way. I see your logic that it shouldn't work when not in browser. But consider for example NVDA+5 keystroke "Report dynamic content change" - it only affects terminal windows, yet it works anywhere. Should this be changed as well? Or should I be consistent with it and make NVDA+8 a global command?
What problem are you trying to solve by proposing to disable NVDA+8 outside of a browser? Do you think new users will be less confused? I don't think so, since users need to understand what does this shortcut do anyway. Same thing applies to all the other NVDA+number row shortcuts.
2. My bad. I'll rename the setting to be camel case.
3. Good idea to use bindGestures(), I'll make the change.

josephsl commented:
Hi, as for item 1, I’d ask for opinions from more experienced people (Jamie, for instance). The NVDA+number row 5 command is something we may need to think about in a future pull request (because it can also affect live regions if needed). Thanks.

 

From: mltony <notifications@github.com> 
Sent: Tuesday, January 1, 2019 11:44 AM
To: nvaccess/nvda <nvda@noreply.github.com>
Cc: Joseph Lee <joseph.lee22590@gmail.com>; Mention <mention@noreply.github.com>
Subject: Re: [nvaccess/nvda] Fast browse mode - "browse mode moves system focus off" mode (#9114)

 

@josephsl <https://github.com/josephsl>  ,

1.	Regarding what should happen if the keystroke is pressed not in browser. I don't htink a strong opinion either way. I see your logic that it shouldn't work when not in browser. But consider for example NVDA+5 keystroke "Report dynamic content change" - it only affects terminal windows, yet it works anywhere. Should this be changed as well? Or should I be consistent with it and make NVDA+8 a global command?
What problem are you trying to solve by proposing to disable NVDA+8 outside of a browser? Do you think new users will be less confused? I don't think so, since users need to understand what does this shortcut do anyway. Same thing applies to all the other NVDA+number row shortcuts.
2.	My bad. I'll rename the setting to be camel case.
3.	Good idea to use bindGestures(), I'll make the change.

—
You are receiving this because you were mentioned.
Reply to this email directly, view it on GitHub <https://github.com/nvaccess/nvda/pull/9114#issuecomment-450752507> , or mute the thread <https://github.com/notifications/unsubscribe-auth/AHgLkN3zrSC1vEJ7parmGdnNAjQAuoYgks5u-7pxgaJpZM4Zltdg> .

mltony committed:
Fix bug when checkboxes and radio button new state is not announced when checked or unchecked

mltony committed:
Renaming configuration setting name to camel case

mltony committed:
Refactoring definitions of passthrough methods

michaelDCurran commented:
The biggest concirn with this approach is web pages with infinite scrolling such as Facebook or Slack. Currently, I believe these pages know to load more data because the first or last article is focused. with browse mode moves system focus off, on these pages, it is dificult to get to new content. At very least, (although this won't fix it)  this pr should not disable scrollIntoView (if it currently does). 
I'd also say that it should still move system focus on quick navigation (if it does not currently). This would at least allow using quick navigation to get to infinite scrolling content. In other words, the only time it should not move focus is with the arrow keys. 
Finally, if we do happen to go with this approach, I do not believe there is a need for the user to toggle this feature. It would always do this.

mltony commented:
@michaelDCurran,
Infinite scrolling webpages is indeed a concern.
1. Calling scrollIntoView causes edit boxes to capture focus on some webpages sporadically.
2. Calling setFocus from within QuickNav commands creates a rubber band effect.
3. I wish we could find a unified way for NVDA to cummunicate with browser that would make everyone happy. However, because of items 1 and 2 I don't see a quick and easy way at this point. So it is a trade off: either we offer full support for infinite scrolling and have users tolerate sluggishness and sporadic focus captures by the edit boxes, or we sacrifice infinite scrolling webpages. Neither option is good enough on its own, so I propose to keep both options at this time and educate users about pros and cons of each mode. The majority of the users don't seem to even notice any sluggishness, so we can make "Browse mode system focus on" to be the default one. If anyone suffers from sluggishness, they can switch to "Browse mode moves system focus off" and suffer the consequences on infinite scrolling pages.

Just to reiterate my points: I don't see a way to eat the cake and have it. There might be a more elaborate solution that would try to autodetect when to sync system focus and when not to, but this will involve a lot more of developing debugging and testing efforts.

mltony commented:
@michaelDCurran, so what is your attitude towards this PR? I know this is not the most elegant solution, but it seems that it's impossible to find an elegant one within current API. Bug 2039 has been open for many years and users are constantly complaining about browse mode problems on NVDA mailing list.

mltony committed:
Refactoring sync focus function

michaelDCurran commented:
@mltony  Firstly thank you for your work on this.
I want to have a bit more of a look into the most common infinite scrolling implementations. If they can be easily identified, I would much rather try and support them, and have browse mode not move system focus the rest of the time. As this really does increase performance. Having a feature that must be turned on won't help the majority of people as they simply won't know to turn it on... and if they did, then it should have been on in the first place.
I will do a bit more investigation and report back here.

jcsteh commented:
Just a reminder that Automatic focus mode for caret movement needs to be
considered here if it hasn't already.

mltony commented:
@michaelDCurran
Do you know if calling scrollIntoView should be enough for infinite scrolling to work? If so, then we can try either calling scrollIntoView only on non-focusable objects, or non-form elements. In any case, I'll try to play with that.

On a separate note, I discovered a couple of bugs with this PR, such as pressing the links in Outlook emails stopped working. I will look into that more, should be easy to fix.

@jcsteh,
I apologize - I'm new to NVDA internals, but what is Automatic focus mode for caret movement ?

mltony committed:
Always call .scrollIntoView()

mltony committed:
Fixing fast browse mode in Outlook emails

mltony committed:
Fixing Tab behavior

mltony committed:
Fixing bug - some edit boxes in IE impossible to enter

mltony committed:
Refactoring

mltony committed:
Refactoring - better tracking of async gainFocus events

mltony commented:
@michaelDCurran
1. I was wrong about .scrollIntoView() causing some side effects - I double checked and it turns out it doesn't cause any problems. So now I call it in fast browse mode (this PR). Also it seems to have fixed infinite scrolling web pages - I only checked Facebook. Now I have more confidence that this PR might be good enough to completely replace current browse mode.
2. I've fixed a few small bugs with this PR

mltony committed:
Fixing exception in event pump

mltony committed:
Rename browse modes

mltony committed:
Set fast browse mode to be default

mltony committed:
Updated bogfixes section

mltony committed:
Add more keystrokes with focus sync

Adriani90 commented:
Bugs also fixed with this PR so far:
* NVDA no longer skips elements on web pages like google translate (#8992)
* NVDA no longer gets stuck in some elements on different web pages (#8999)
* Words are no longer being cut while reporting and elements are no longer mixed with each other while jumping from element to element in browse mode (#9129)

michaelDCurran commented:
I will review the code hopefully early next week.

I am concerned about your statement that you received "mostly" positive 
feedback. I guess there still are some corner cases that don't work well?

If I think the code is okay, perhaps we can get it into the next 
release, off by default, but let people know how to test out the new 
experimental fast browseMode.

Then perhaps in the following release we might be ready to turn it on by 
default.

How does that sound?

dingpengyu commented:
@michaelDCurran 
Since there is no big problem with the code, why not fix the game error and merge it once? Flying to the next version?

ABuffEr commented:
Hi,
I'm one of users tony said :)
I'm trying again the compiled build provided for testing, and I confirm that there is a refresh bug that I see with notifications button on Facebook (desktop site, with Firefox 52.9.0 ESR 64-bit, Windows 10 64-bit). I'll try to explain a bit more...
On the top of Facebook page there are some buttons, Requests, Messages, Notifications. When a new notification arrives (for a comment or a like, usually), the label of Notifications button becomes "1 Notifications". Then, when you enter on button, read or mark as read notification and close the dialog with escape, button label returns to Notifications.
Well, with this testing build, I must force refresh with NVDA+F5 to have first "1 Notifications" and then, after reading, "Notification" label again. If it's useful, this is the log of button (label is in Italian):
***
INFO - globalCommands.GlobalCommands.script_navigatorObject_devInfo (13:00:21.094):
Developer info for navigator object:
name: u'Notifiche'
role: ROLE_BUTTON
states: STATE_FOCUSABLE, STATE_FOCUSED
isFocusable: True
hasFocus: True
Python object: <NVDAObjects.Dynamic_ButtonMozillaIAccessible object at 0x061EF8F0>
Python class mro: (<class 'NVDAObjects.Dynamic_ButtonMozillaIAccessible'>, <class 'NVDAObjects.IAccessible.Button'>, <class 'NVDAObjects.IAccessible.mozilla.Mozilla'>, <class 'NVDAObjects.IAccessible.ia2Web.Ia2Web'>, <class 'NVDAObjects.IAccessible.IAccessible'>, <class 'NVDAObjects.window.Window'>, <class 'NVDAObjects.NVDAObject'>, <class 'documentBase.TextContainerObject'>, <class 'baseObject.ScriptableObject'>, <class 'baseObject.AutoPropertyObject'>, <type 'object'>)
description: u''
location: RectLTWH(left=1535, top=128, width=48, height=48)
value: None
appModule: <'firefox' (appName u'firefox', process ID 18176) at address 5d7af70>
appModule.productName: u'Firefox'
appModule.productVersion: u'52.9.0'
TextInfo: <class 'NVDAObjects.IAccessible.IA2TextTextInfo'>
windowHandle: 2559004L
windowClassName: u'MozillaWindowClass'
windowControlID: 0
windowStyle: 399441920
windowThreadID: 15264
windowText: u'Facebook - Mozilla Firefox'
displayText: u''
IAccessibleObject: <POINTER(IAccessible2) ptr=0x7f952fc at 6172620>
IAccessibleChildID: 0
IAccessible event parameters: windowHandle=2559004, objectID=-4, childID=-15057
IAccessible accName: u'Notifiche'
IAccessible accRole: ROLE_SYSTEM_PUSHBUTTON
IAccessible accState: STATE_SYSTEM_FOCUSED, STATE_SYSTEM_FOCUSABLE, STATE_SYSTEM_VALID (1048580)
IAccessible accDescription: u''
IAccessible accValue: None
IAccessible2 windowHandle: 2559004
IAccessible2 uniqueID: -15057
IAccessible2 role: ROLE_SYSTEM_PUSHBUTTON
IAccessible2 states: IA2_STATE_SELECTABLE_TEXT, IA2_STATE_OPAQUE (5120)
IAccessible2 attributes: u'margin-left:2px;text-align:start;draggable:true;text-indent:0px;formatting:block;margin-right:2px;xml-roles:button;tag:a;class:jewelButton _3eo9 f_key;margin-top:0px;margin-bottom:0px;display:block;explicit-name:true;'
***
The other bug about refreshing on "Other..." button to expand post content (sorry, long to explain if you don't know Facebook) is probably caused by shift+enter sync lackness, that I see from here Tony has resolved, so do not consider for the moment :)

Adriani90 commented:
In Firefox, shift+enter is used by default to open a link in a new window. This works not as expected in fast browse mode because the focus is not moving. I know some users who use this key stroke in Firefox. So I think at least we should document properly that the key stroke does not work in Firefox when using fast browse mode. At least this was the behavior in the snapshot_source-fast_nvda-7c9d81a.

Adriani90 commented:
@leonardder could you please test if it works as expected from a braille display user point of view? Thanks!

mltony commented:
@ABuffEr,
Yes, I could reproduce this problem on my computer. Using latest Firefox and Windows 10. I was sending myself messages from another Facebook account and observing whether the button will be updated. Here is what I found:
1. In Firefox in fast browse mode the notification button is indeed not updated when a new message arrives unless you press NVDA+F5.
2. In traditional browse mode when you just navigate to this button, at first it says "no new messages", and then a split second later it says "1 new message". Which might suggest  that the website probably detects .setFocus() call and only then updates the title of the button for the screenreaders.
3. The same problem is also present in JAWS - which might indicate that JAWS doesn't call .setFocus on every keystroke. After pressing Insert+Escape (buffer refresh), it correctly sees updated "1 messages" notification.
4. I couldn't reproduce this problem in Chrome, IE and Edge - notification is updated correctly and automatically in fast browse mode in NVDA.
So here are my conclusions:
1. It is probably impossible to fix this problem in Fast browse mode, since the main idea of fast browse mode is not to call .setFocus on every keystroke.
2. This is exactly the reason why I prefer to keep traditional browse mode around. There are probably more web sites that rely on .setFocus() call being called and there are NVDA users relying on this behavior.
3. I would argue that his issue is not a good reason not to go ahead with fast browse mode. Fast browse mode fixes a number of severe problems with browse mode. So maybe a reasonable compromise would be to keep both traditional and fast browse modes together and let the users to switch between them.

@Adriani90,
I have already fixed Shift+Enter in this PR, I haven't made  a new build though as this fix was very straightforward.

ABuffEr commented:
Hi Tony,
I return on this issue only to say that, according to your observations, moving focus with NVDA+shift+backspace (from another item to Notifications) seems to get same result of refreshing, with advantage of keep current position.
Sure, the best solution could be to know whether current HTML item needs to get focus, but it's probably hard or impossible to find from a screenreader perspective, I imagine...
So, workaround of NVDA+shift+backspace could be an acceptable solution.

mltony commented:
@ABuffEr,
Are you using some add-on for NVDA+Shift+Backspace keystroke? It doesn't have any NVDA action  assigned to it on my computer.
There are multiple workarounds available:
1. Switch back to traditional browse mode.
2. Press NVDA+F5 to refresh the virtual buffer.
3. Not sure about NVDA+Shift+Backspace, but it is definitely possible to create a new keystroke that would manually route system focus to browse focus on demand.
4. And finally, file an accessibility bug with Facebook.

Adriani90 commented:
Hey @mltony, the key stroke nvda+shift+backspace moves the Focus to the nvda Cursor in Laptop Keyboard layout.

mltony commented:
@michaelDCurran, kindly reminder - did you have a chance to look at this PR? Thanks!

mltony committed:
Renaming legacy browse mode back to traditional browse mode since we won't remove it in the nearby future

mltony commented:
@michaelDCurran, do you have any estimate when this could be reviewed? Do we have enough time to squeeze this PR into 2019.1 release?

michaelDCurran commented on source/globalCommands.py:
```diff
@@ -1534,6 +1534,20 @@ def script_toggleFocusMovesNavigatorObject(self,gesture):
+			config.conf["virtualBuffers"]["focusFollowsBrowse"]=False
+		else:
+			# Translators: presented when toggled.
+			state = _("Traditional browse mode")
```
I would prefer that these messages reflected similar wording to what is in the Settings dialog. I.e. Focus following browse mode, or focus not following browse mode.   "Fast / traditional" is a little mis-leading I think :)

michaelDCurran commented on source/browseMode.py:
```diff
@@ -250,6 +250,13 @@ def _getLabelForProperties(self, labelPropertyGetter):
 
+def getObjectID(obj):
+	try:
+		return obj.IA2UniqueID
```
NVDAObjects should have a new 'uniqueID' property. For IA2 it should be self.windowHandle,self.ia2UniqueID, for UIA it should be self.UIAElement.getRuntimeId(). For generic IAccessible NVDAObjects it should be self.IAccessibleIdentity if it exists otherwise hash(self.IAccessibleObject). For Window it should be self.windowHandle.

mltony replied:
I just checked current head branch - NVDA objects don't have uniqueID property. Do you mean it will be added in the future or am I missing something? the In any case, this code is only needed to make sure that async part of code is called on the same NVDA object - which in all cases that I know of it is. So it is kind of a second line of defense - to identify async-related bugs early.

michaelDCurran replied:
I was requesting that you implement the uniqueID property on NVDAObjects as I described. If you fully trust this works without this check, then remove it. However, I cannot approve the code as is with specific mention to IAccessible2 in browseMode.py.

mltony replied:
Sorry for misunderstanding - I implemented .uniqueID property as you requested in IAccessible, UAI and Window classes. I couldn't find "generic IAccessible NVDAObjects " - could you point me where it is located? But now browseMode doesn't have any IA2-specific code any more.

michaelDCurran commented on source/browseMode.py:
```diff
@@ -477,10 +488,51 @@ def _activatePosition(self,obj=None):
 
 	def script_activatePosition(self,gesture):
-		self._activatePosition()
+		self.maybeSyncFocus(postFocusFunc=self._activatePosition)
```
Can you explain what would happen if activatePosition was executed straight away possibly before setfocus completes? I.e. why the need for the asynchronis stuff?

mltony replied:
Good question - I expected you to ask it :-)

The problem I was facing was that when you click on a checkbox on a web page, its updated status won't be announced. This also applies to radio buttons and potentially other form controls.

Here are what my investigation has shown:
Typically in NVDA speech.speakObjectProperties is called twice:
* First, when a checkbox is focused, e.g. browse mode cursor is on it. It is called with reason=controlTypes.REASON_ONLYCACHE, which doesn't actuallyspeak, but only remembers its state. 
* Second, when script_activatePosition() is called,, e.g. user pressed enter on a checkbox, speakObjectProperties is called with reason="change", which compares all the properties of NVDAObject and only speaks whatever is changed.

With FastBrowseMode, Focus event and activate event are issued very closely together. Since both of them are processed asynchronously, via event dispatcher, focus processing (browsermode.event_gainFocus()) is actually called after activatePosition(), because it is called asynchronously and naturally there is some delay. As a result, it sees already updated state of checkbox. As a result, the next call to speak does nothing, because from its point of view nothing has changed.

The way to resolve this problem is to make sure that focus event and all the postprocessing happens stricly before activatePosition call is issued. So we have to wait until async event onFocusGain  comes in and is fully processed.

michaelDCurran commented on source/browseMode.py:
```diff
@@ -765,6 +817,15 @@ def script_disablePassThrough(self, gesture):
 	prevError=_("no previous error"))
 del qn
 
+PT = BrowseModeTreeInterceptor.addPassThrough
```
Why don't you just add these to BrowseModeTreeInterceptor.__gestures?

mltony replied:
Just added. Joseph wanted me to write it this way.

michaelDCurran commented on source/browseMode.py:
```diff
@@ -477,10 +488,51 @@ def _activatePosition(self,obj=None):
+	def maybeSyncFocusAndPassThrough(self, gesture=None): 		
+		self.maybeSyncFocus()
+		if gesture:
+			gesture.send()
```
I'm not sure if this method is necessary. I think it is more readable to keep the gesture.send in the original places outside of this method.

michaelDCurran reviewed (changes requested):
The userGuide must also be updated in this pr.

mltony committed:
Update user documentation

mltony committed:
Change toggle messages

mltony committed:
Use .uniqueID attribute

mltony committed:
Revert back to IA2UniqueID

mltony committed:
Refactoring

mltony committed:
Refactoring

mltony committed:
Refactoring

mltony committed:
Refactoring

dnz3d4c commented:
Could this pr be merged into NVDA 2019.1?

michaelDCurran commented:
@dnz3d4c: Could you please refrain from asking if PRs can be merged?

When / if a PR is approved it will be merged.

Thank you.

dingpengyu commented:
> @dnz3d4c: Could you please refrain from asking if PRs can be merged? When / if a PR is approved it will be merged. Thank you.

I hope I can see this PR in NVDA2019.1.
@mltony 
@michaelDCurran 
@Adriani90 
@josephsl

Adriani90 commented:
@dnz3d4c, @dingpengyu I hope you read also the comments of pull requests and not only the description. There were already several People asking this PR to be included in 2019.1. People are working on it and as soon as it is aproved it will be merged. If you want to express your wilingness to have it in 2019.1, please do so in the NVDA user list on Groups.io. Main developers incl. Quentin from NV Access are there as well and will notice your wishes. But please don't spam Pull requests with comments like this. If you can contribute constructively through testing of try builds etc, your comments are very welcome on the pull requests. But otherwise please use the user list. Thanks.

mltony committed:
Adding .uniqueID attribute to NVDAObject implementations

mltony committed:
Adding passthrough gestures with numpadEnter

michaelDCurran commented on source/NVDAObjects/IAccessible/__init__.py:
```diff
@@ -627,6 +627,7 @@ def __init__(self,windowHandle=None,IAccessibleObject=None,IAccessibleChildID=No
 		self.event_objectID=event_objectID
 		self.event_childID=event_childID
 		super(IAccessible,self).__init__(windowHandle=windowHandle)
+		self.uniqueID = (self.windowHandle, self.IA2UniqueID)
```
It would be better if this was a lazy property rather than set in the constructor.
The Generic IAccessible code I'm talking about can also go here.
E.g.:
```
def _get_uniqueID(self):
	if self.ia2UniqueID:
		return (self.windowHandle,self.ia2UniqueID)
	elif self.IAccessibleIdentity:
		return self.IAccessibleIdentity
	else:
		return (self.windowHandle,hash(self.IAccessibleObject)
```

michaelDCurran commented on source/NVDAObjects/UIA/__init__.py:
```diff
@@ -902,6 +902,7 @@ def __init__(self,windowHandle=None,UIAElement=None,initialUIACachedPropertyIDs=
 		if not windowHandle:
 			raise InvalidNVDAObject("no windowHandle")
 		super(UIA,self).__init__(windowHandle=windowHandle)
+		self.uniqueID =         self.UIAElement.getRuntimeId()
```
Like with IAccessible, it would be better if this were a lazy property.

mltony committed:
Lazy evaluation of uniqueID

michaelDCurran commented on source/browseMode.py:
```diff
@@ -501,7 +501,7 @@ def maybeSyncFocus(self, postFocusFunc=None):
 
 		if not synchronousCall:
-			uid = obj.uniqueID
+			uid = obj._get_uniqueID()
```
obj.uniqueID should still work here. the ```_get_*``` methods are magic properties :)

mltony replied:
Got it. So many fancy features in Python!

mltony committed:
Refactoring

michaelDCurran committed:
Merge branch 'master' into HEAD

michaelDCurran commented:
@josephsl Do you still want changes on this, or are you happy to approve?

josephsl commented:
Hi,

I personally think it should be tabled for 2019.2 unless people say it can board the beta train. As for final things, I'll get back on that once I run it through tests tonight.

Thanks.

josephsl commented:
Hi,

Signing off.

To @mltony: congratulations on your first major project. However, the work is just beginning - keep an eye on feedback from more people.

Thanks.

josephsl commented:
Hi,

As for my comment on tabling this feature: please ignore it. Thanks.

mltony commented:
@michaelDCurran, 
Thank you for review!
@josephsl, @Adriani90, @ABuffEr
Thank you guys too.
I'll definitely keep an eye on any related bug reports. I wouldn't be working on this PR if I knew I couldn't maintain it.

gregjozk commented:
> @jcsteh, I apologize - I'm new to NVDA internals, but what is Automatic focus mode for caret movement ?

if enabled ( from UserGuide:

> This option, when checked, allows NVDA to enter and leave focus mode when using arrow keys. For example, if arrowing down a web page and you land on an edit box, NVDA will automatically bring you into focus mode. If you arrow out of the edit box, NVDA will put you back in browse mode.

sadly this function is partially broken with focus off.

str:
- turn off focus (nvda+8) in FF for example
- go to Google.com
- with arrow keys go to edit field
- try with arrow keys to exit from it

i could not exit from the edit field with arrow keys; had to press other keys to go back to browse mode.

Adriani90 commented:
Cc: @mltony

Von meinem iPhone gesendet

> Am 20.02.2019 um 02:41 schrieb gregjozk <notifications@github.com>:
> 
> @jcsteh, I apologize - I'm new to NVDA internals, but what is Automatic focus mode for caret movement ?
> 
> if enabled ( from UserGuide:
> 
> This option, when checked, allows NVDA to enter and leave focus mode when using arrow keys. For example, if arrowing down a web page and you land on an edit box, NVDA will automatically bring you into focus mode. If you arrow out of the edit box, NVDA will put you back in browse mode.
> 
> sadly this function is partially broken with focus off.
> 
> str:
> 
> turn off focus (nvda+8) in FF for example
> go to Google.com
> with arrow keys go to edit field
> try with arrow keys to exit from it
> i could not exit from the edit field with arrow keys; had to press other keys to go back to browse mode.
> 
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub, or mute the thread.

gregjozk commented:
Hi,

after testing I understand, why it is "broken". Maybe one line in UG
shell explain it also for others:

e.g.: due to technical reasons function "automatic focus mode for
carret movement" does not work as expected when fast browse mode is on
( focus follows browse mode off).

regards, Jožef

2019-02-20 3:44 GMT+01.00, Adriani90 <notifications@github.com>:
> Cc: @mltony

>

> Von meinem iPhone gesendet

>

>> Am 20.02.2019 um 02:41 schrieb gregjozk <notifications@github.com>:

>>

>> @jcsteh, I apologize - I'm new to NVDA internals, but what is Automatic
>> focus mode for caret movement ?

>>

>> if enabled ( from UserGuide:

>>

>> This option, when checked, allows NVDA to enter and leave focus mode when
>> using arrow keys. For example, if arrowing down a web page and you land on
>> an edit box, NVDA will automatically bring you into focus mode. If you
>> arrow out of the edit box, NVDA will put you back in browse mode.

>>

>> sadly this function is partially broken with focus off.

>>

>> str:

>>

>> turn off focus (nvda+8) in FF for example

>> go to Google.com

>> with arrow keys go to edit field

>> try with arrow keys to exit from it

>> i could not exit from the edit field with arrow keys; had to press other
>> keys to go back to browse mode.

>>

>> —

>> You are receiving this because you were mentioned.

>> Reply to this email directly, view it on GitHub, or mute the thread.

>

>

> --

> You are receiving this because you commented.

> Reply to this email directly or view it on GitHub:

> https://github.com/nvaccess/nvda/pull/9114#issuecomment-465397861

ehollig commented:
Should this have closed #2039?

mltony commented:
@gregjozk,
What browser / OS are you using?
I did a quick test of interference of fast browse mode with "Automatic focus mode for caret movement " on my computer, and here are results:
1. Google Chrom - works as expected
2. Firefox - works as expected
3. IE - broken in fast browse mode, but works fine in traditional browse mode
4. Edge - broken in both fast and  traditional browse mode.
I'll take alook on the weekend why it's broken in IE. But judging by the fact that this feature doesn't work in edge to begin with - it's not being widely used.

This PR is supposed to close many open bugs. However I wouldn't close them just yet - before making sure that fast browse mode doesn't have any major regressions.

gregjozk commented:
Hi,

tested with FF 60 ESR and Chrome on w10 (1809)
but as I said, I understand why it works as it works;  and I'm glad,
because modern websites like FB and others are more readable.

thx.

p.s. when you have tested on Google.com did you disable automatic
suggestions? as I observed it has not work on edit fields, where
automatic suggestions were disabled. strange, I knoe.

2019-02-20 18:49 GMT+01.00, mltony <notifications@github.com>:
> @gregjozk,
> What browser / OS are you using?
> I did a quick test of interference of fast browse mode with "Automatic focus
> mode for caret movement " on my computer, and here are results:
> 1. Google Chrom - works as expected
> 2. Firefox - works as expected
> 3. IE - broken in fast browse mode, but works fine in traditional browse
> mode
> 4. Edge - broken in both fast and  traditional browse mode.
> I'll take alook on the weekend why it's broken in IE. But judging by the
> fact that this feature doesn't work in edge to begin with - it's not being
> widely used.
>
> This PR is supposed to close many open bugs. However I wouldn't close them
> just yet - before making sure that fast browse mode doesn't have any major
> regressions.
>
>
>
> --
> You are receiving this because you were mentioned.
> Reply to this email directly or view it on GitHub:
> https://github.com/nvaccess/nvda/pull/9114#issuecomment-465684525

mltony commented:
Ok, I could reproduce this issue in Chrome on a plain HTML edit box on this page:
https://www.w3schools.com/html/html_forms.asp
I'll take a look why this is happening.

ABuffEr commented:
Hi,
I cannot provide an exact steps-to-reproduce, but I noticed some discrepances when I open a web dialog and then interact with an item pressing space.
Then, but I'm not sure, I have the impression that some edit fields are not focused/labelled correctly, maybe when are within sections...
As usual, my tests come mainly from Firefox and Facebook (notifications web dialog, and comment/reply fields), where we know system focus has its importance. So, let be free to ignore it :)
In addition, a user on Italian mailing list asks whether it's possible to use NVDA+tab to know current position, and not current focus position.

ABuffEr commented:
Hi again,
sorry, I'm forgetting... a question: actually, can switching between fast and traditional mode and vice versa be considered stable? I mean: it can (should) be done without restarting NVDA or browser?
To use traditional mode on Facebook, and fast mode in other sites, for example...
Thanks.

jcsteh commented:
I'd suggest that forcing focus mode with NVDA+space should set focus to the nearest focusable object under the cursor. There are some cases where you want to use the web app's native keyboard shortcuts, but enter wouldn't activate focus mode and those shortcuts operate relative to the focus; e.g. Twitter.

michaelDCurran commented:
Just a heads up that this pr may have to be reverted for 2019.1 as we are already behind on releasing a beta, and there are at least 2 issues reported due to this being merged. Having said this though, I have seen positive feedback, and I would still love the feature to be fixed up and merged in the next release.

mltony commented:
@michaelDCurran,
So here is the current status of all known issues.
1. Facebook notifications button doesn't get updated autmatically in Firefox - impossible to fix - as I described above, since it is only updated when focus event is sent to it.
2. NVDA+Tab doesn't announce correct element - I fixed that, haven't sent PR yet.
3. Auto forms mode doesn't work - I fixed that - haven't sent PR yet.
4. NVDA+Insert doesn't activate current element - haven't fixed it yet, but should be a quick fix. Most likely I'll be able to fix it and send you a PR this coming weekend, if that's not too late.
Now about the last and very tricky issue - about async call. I already described to you above why I needed this async call. Now however, it turns out there are too many problems with it. I suspect in issue #9320 the problem is that we get gainFocus call for a different object that we called setFocus on. This alone couldn've been probably fixed. However, there seems to be another problem, where in approximately 0.01%  of cases gainFocus event just doesn't come at all. I have no idea why this is happening, and I don't see a good way to debug that. 
I had my own good reasons to use async calls. In theory it seems possible to implement this without async calls. But, I would have to first get the logic of gainFocus event and run it right after I call setFOcus synchronously. Moreover, I'd have to hijack the behavior of api.setFocusedObject and api.getFocusObject() calls, as NVDA's current focused object only changes when gainFocus event is received, and if we don't wait for callback, we might have to provide to other functions currently focused object in another way. In other words, without async call, right after we call .setFocus the value of api.getFocuesed Object is going to be invalid. So without async call, much more code changes will need to be made.
Now having said all of that, it is still your call whether to roll back this PR or not. I have been testing this PR for almost two months already and it works beautifully for me, except for this 0.01% chance of exception. If I start working on reverting async calls - this might take me quite some time - I have a day job and I only have limited time on the weekends to work on this. If you ask me for my opinion, I would say that this PR is going to be useful to people even despite the async bug. But again, it's your call. What I can tell you at this point, is that I don't have confidence that I will be able to make it to work without async call, as a lot of time and effort will be required. I may give it a try though. And with async call it seems like there are some tiny bugs that are nearly impossible to fix.

michaelDCurran commented:
I made the decision to revert this pr, and have done so by merging revert pr #9327.
I look forward to seeing this polished further and re-submitted for 2019.2.

ABuffEr commented:
Hi,
maybe for optimization, or not... I often get this error, quite annoying in snapshot (even if I use traditional browse mode, so with latest alpha too):
ERROR - eventHandler.executeEvent (18:10:47.924):
error executing event: caret on <NVDAObjects.IAccessible.mozilla.Mozilla object at 0x057DB590> with extra args of {}
Traceback (most recent call last):
  File "eventHandler.pyc", line 155, in executeEvent
  File "eventHandler.pyc", line 92, in __init__
  File "eventHandler.pyc", line 98, in next
  File "eventHandler.pyc", line 126, in gen
  File "NVDAObjects\__init__.pyc", line 361, in _get_treeInterceptor
  File "treeInterceptorHandler.pyc", line 20, in getTreeInterceptor
  File "virtualBuffers\gecko_ia2.pyc", line 143, in __contains__
  File "browseMode.pyc", line 1571, in _isNVDAObjectInApplication
  File "virtualBuffers\__init__.pyc", line 702, in _isNVDAObjectInApplication_noWalk
  File "virtualBuffers\gecko_ia2.pyc", line 177, in getIdentifierFromNVDAObject
AttributeError: 'Desktop' object has no attribute 'IA2UniqueID'

mltony commented:
@michaelDCurran,
I understand your decision. But I would like to know, what is your policy towards bugs in PRs? Basically, as I mentioned there are bugs/side effects in this PR that are impossible to fix, such as Facebook notifications. There is another set of bugs related to async behavior, that are very hard to fix. Do you have policy of accepting only no bugs PRs? In this case this PR has no chances. Or you'd be willing to accept PR with some known issues?

MarcoZehe commented:
Just my 2 Cents:

>     1. Facebook notifications button doesn't get updated autmatically in Firefox - impossible to fix - as I described above, since it is only updated when focus event is sent to it.

This *will* be viewed by the average user as a regression from a previous NVDA version, and they're not going to like it especially if they are heavy Facebook users. And for that perception, it doesn't matter what technical reason there might be for it. All these users will see is "there is this change in NVDA that supposedly makes things better", but subjectively, it breaks functionality for them they might have come to rely on for months or even years. So for them, this is no improvement at all, despite the advantage this might have elsewhere and not touch any of their use cases at all. For them, this is a deal breaker. I realize that many NVDA users might be more forgiving than those who use a paid screen reader, but the more non-high-tech users NVDA gains, the bigger the number of users will be who will not tolerate such an obvious regression.

Adriani90 commented:
@MarcoZehe I agree in most with you, however in my opinion we could call it (experimental) and it could be moved to advanced settings like UIA in MS Word for example which also break things when is used. I think such features could be indeed merged in NVDA but the user should be aware that it is an experimental feature, maybe just in brackets on the label of the checkbox. But in any case bugs related to that experimental feature should not get forgotten for a long time. At least for me, the browser works actually quite stable even with focus not following browse mode. There are however some sites where it does not work but I know that it is an experimental feature and I can turn it off.

Adriani90 commented:
And this is in my opinion the most complicated aspect to decide on. Do we want many experimental features to be merged and who knows when the bugs get solved? Or do we want such experimental features to stay in open PRs and to be available only through try builds?
In my opinion, if I were a developer, I would try to implement experimental features from time to time but not too much of them. This could be like an indirect survey where users would find bugs and report them to you. And then you can really have a powerful feature which has been widely tested and you are sure that bugs are unlikely to happen if the reported bugs can be solved.

In this specific case, there are bugs which seem impossible to solve. So the question is, are there any other alternatives to solve the bugs referenced above which in fact are solved through this PR without causing additional bugs? From this point of view, I can understand @michaelDCurran's decision as well. But if there is no alternative, then I think most people would accept smaller bugs while more annoying bugs get solved. And this is the case with this PR. There are some small bugs caused by it, but it solves some really serious bugs such as cursor getting stuck in a region, buttons not being reported correctly etc.

michaelDCurran commented:
@mltony: I'm afraid there is no hard and fast rule here. To some degree 
I need to use my personal judgement for these decisions. 2019.1 is 
already very delayed (due to other reasons) but the first beta will be 
ready to go in the next day or so. I am not convinced that those bugs 
cannot be addressed. I think we can spend more time on it for the next 
release. And again, just because it is backed out of 2019.1 does not 
mean that it won't get into 2019.2.

At the end of the day though, NV Access is primarily responsible for 
anything we put in a release, and right now I did not think it was a 
good fit for 2019.1.

mltony commented:
@michaelDCurran,
I totally understand reverting this PR for 2019.1. I will try to fix as many bugs as possible and come back with updated PR for 2019.2. Getting rid of async calls might be a challenge, so I will discuss this with you if I won't be able to do it myself.

@MarcoZehe
That's a very defensive attitude, and it is hard to argue with. You can claim that we cannot tolerate any regressions here, including the Facebook notification button. But that would imply that a dozen browse mode bugs mentioned above will still be there, some of them open for 7 years now.
To me it seems that sending .setFocus on every cursor move was a bug in the first place (no offense to NVDA devs, it was probably a reasonable choice bakc then, but then again, there's a dozen open issues). And as it often happens, people tend to rely on certain behaviors, even buggy ones. And I don't mean NVDA users here, I mean Facebook UI developers who rely on NVDA sending setFocus event to update notifications count.
So it's a trade-off, it's a questions of whether we want to move forward.

jcsteh commented:
Regarding the Facebook Notifications button, JAWS never set focus to
buttons like NVDA does. So surely JAWS users have experienced this for
years? Does this work in Chrome? Is this a firefox specific issue somehow?

mltony commented:
@jcsteh,
Yes, according to my tests, this issue only appears in Firefox. Other browsers (Chrome, IE, Edge) work as expected.
Jaws behavior is exactly the same - issue can be reproduced in Firefox, but works as expected in all other browsers.
So based on this observation it is consistent that Jaws doesn't call setFocus.
I don't think this is Firefox specific, I rather think that FB might be running some Firefox specific code that triggers this issue.

jcsteh commented:
There's another possibility, which is that Firefox isn't firing the right
event for some reason when that button changes. @MarcoZehe, can you please
look into this when you're able?
